java/ql/src/Stubs/Stubs.qll

```diff
@@ -0,0 +1,440 @@
+/**
+ * Generates java stubs for use in test code.
+ *
+ * Extend the abstract class `GeneratedDeclaration` with the declarations that should be generated.
+ * This will generate stubs for all the required dependencies as well.
+ */
+
+import java
+
+/** A type that should be in the generated code. */
+abstract private class GeneratedType extends RefType {
+  GeneratedType() {
+    (
+      this instanceof Interface
+      or
+      this instanceof Class
+    ) and
+    not this instanceof AnonymousClass and
+    not this instanceof LocalClass and
+    not this.getPackage() instanceof ExcludedPackage
+  }
+
+  private string stubKeyword() {
+    this instanceof Interface and result = "interface"
+    or
+    this instanceof Class and result = "class" and not this instanceof EnumType
+    or
+    this instanceof EnumType and result = "enum"
+  }
+
+  private string stubAbstractModifier() {
+    if this.(Class).isAbstract() then result = "abstract " else result = ""
+  }
+
+  private string stubStaticModifier() {
+    if this.isStatic() then result = "static " else result = ""
+  }
+
+  private string stubAccessibilityModifier() {
+    if this.isPublic() then result = "public " else result = ""
+  }
+
+  /** Gets the entire Java stub code for this type. */
+  final string getStub() {
+    result =
+      this.stubAbstractModifier() + this.stubStaticModifier() + this.stubAccessibilityModifier() +
+        this.stubKeyword() + " " + this.getName() + stubGenericArguments(this) +
+        stubBaseTypesString() + "\n{\n" + stubMembers() + "}"
+  }
+
+  private RefType getAnInterestingBaseType() {
+    result = this.getASupertype() and
+    not result instanceof TypeObject and
+    not this instanceof EnumType and
+    result.getSourceDeclaration() != this
+  }
+
+  private string stubBaseTypesString() {
+    if exists(getAnInterestingBaseType())
+    then
+      exists(string cls, string interface, string int_kw | result = cls + int_kw + interface |
+        (
+          if exists(getAnInterestingBaseType().(Class))
```

Comment:
Check: what about an interface extending a superinterface?

Reply:
This should be handled by the next part of this predicate